index.js

```diff
@@ -1,42 +1,51 @@
 'use strict';
 
+var isArray = Array.isArray;
+var keyList = Object.keys;
+var hasProp = Object.prototype.hasOwnProperty;
+
 module.exports = function equal(a, b) {
   if (a === b) return true;
 
-  var arrA = Array.isArray(a)
-    , arrB = Array.isArray(b)
-    , i;
+  var arrA = isArray(a)
+    , arrB = isArray(b)
+    , i
+    , length
+    , key;
 
   if (arrA && arrB) {
-    if (a.length != b.length) return false;
-    for (i = 0; i < a.length; i++)
+    length = a.length;
+    if (length != b.length) return false;
+    for (i = 0; i < length; i++)
       if (!equal(a[i], b[i])) return false;
     return true;
   }
 
   if (arrA != arrB) return false;
 
-  if (a && b && typeof a === 'object' && typeof b === 'object') {
-    var keys = Object.keys(a);
-    if (keys.length !== Object.keys(b).length) return false;
+  var dateA = a instanceof Date
+    , dateB = b instanceof Date;
+  if (dateA != dateB) return false;
+  if (dateA && dateB) return a.getTime() == b.getTime();
 
-    var dateA = a instanceof Date
-      , dateB = b instanceof Date;
-    if (dateA && dateB) return a.getTime() == b.getTime();
-    if (dateA != dateB) return false;
+  var regexpA = a instanceof RegExp
+    , regexpB = b instanceof RegExp;
+  if (regexpA != regexpB) return false;
+  if (regexpA && regexpB) return a.toString() == b.toString();
 
-    var regexpA = a instanceof RegExp
-      , regexpB = b instanceof RegExp;
-    if (regexpA && regexpB) return a.toString() == b.toString();
-    if (regexpA != regexpB) return false;
+  if (a instanceof Object && b instanceof Object) {
+    var keys = keyList(a);
+    length = keys.length;
 
-    for (i = 0; i < keys.length; i++)
-      if (!Object.prototype.hasOwnProperty.call(b, keys[i])) return false;
+    for (i = 0; i < length; i++)
+      if (!hasProp.call(b, keys[i])) return false;
 
-    for (i = 0; i < keys.length; i++)
-      if(!equal(a[keys[i]], b[keys[i]])) return false;
+    for (i = 0; i < length; i++) {
+      key = keys[i];
+      if (!equal(a[key], b[key])) return false;
+    }
 
-    return true;
+    return length == keyList(b).length;
```

Comment:
sorry, I can be missing something, but shouldn't this condition be on line 39 as `if (length != keyList(b).length) return false`? It looks like the simplest and the fastest inequality check - if the number of keys is different it should immediately return false, rather than comparing all key values first.

Reply:
I think `keyList(b).length` is very expensive, and when moving it to the top results drop.

The reason is that `keyList` creates a new array, which we postpone to the very end, only if necessary.

For, example if `b` is

```
{
  a: 1,
  b: 1,
  b: 1,
}
```

`keyList` has to create an array for us

```js
['a', 'b', 'c']
```

just so we can look at its `.length` property.

I am actually thinking now, maybe I should iterate through the object myself and count the number of keys in each object instead of doing `keyList(a).length` and `keyList(b).length`.

Reply:
I think the results will depend on the size of the object. Checking and comparing the number of keys can be expensive, but it's predictably expensive and should scale linearly with the number of keys. Comparing the values in keys can be unpredictably slow as they can have very deep data hierarchies. So I think we should compare the number of keys as the first thing and then either iterate and compare key names (as is now) or it may be faster to call equal(keysArray1, keysArray2) - not sure, probably extra function call will make slower.